File: fXDKit/Classes/_ObjC/_module/FXDmoduleMessage.m


#import "FXDmoduleMessage.h"


@implementation FXDmoduleMessage

#pragma mark - Memory management

#pragma mark - Initialization

#pragma mark - Property overriding

#pragma mark - Method overriding

#pragma mark - Public
- (void)presentEmailScene:(MFMailComposeViewController*)emailScene forPresentingScene:(UIViewController*)presentingController usingImage:(UIImage*)image usingMessage:(NSString*)message withRecipients:(NSArray*)recipients {	FXDLog_DEFAULT;


	if ([MFMailComposeViewController canSendMail] == NO) {
		//FIXME: alert user
		return;
	}


	if (emailScene == nil) {
		if (image || message) {
			emailScene = [self emailSceneForSharingImage:image usingMessage:message];
		}
		else {
			emailScene = [self emailSceneWithMailBodyWithRecipients:recipients];
		}
	}

	if (presentingController == nil) {
		FXDWindow *mainWindow = (FXDWindow*)[UIApplication sharedApplication].delegate.window;

		if (mainWindow.rootViewController) {
			FXDLogObject(mainWindow.rootViewController);

			presentingController = mainWindow.rootViewController;
		}
	}


	emailScene.mailComposeDelegate = self;

	[presentingController
	 presentViewController:emailScene
	 animated:YES
	 completion:nil];
}

- (MFMailComposeViewController*)emailSceneWithMailBodyWithRecipients:(NSArray*)recipients {	FXDLog_DEFAULT;
	FXDLogObject([[NSBundle mainBundle] infoDictionary]);
	FXDLogObject(recipients);

	NSString *version = application_BundleVersion;
	FXDLogObject(version);

#if ForDEVELOPER
	NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
	NSArray *languages = [userDefaults objectForKey:@"AppleLanguages"];
	NSString *currentLanguage = languages.firstObject;
	FXDLogObject(currentLanguage);
#endif


	NSString *displayName = application_DisplayName;

	NSString *subjectLine = [NSString stringWithFormat:@"[%@]", displayName];

	NSString *lineSeparator = @"_______________________________";
	NSString *appVersionLine = [NSString stringWithFormat:@"%@ %@", subjectLine, version];


	struct utsname systemInfo;
	uname(&systemInfo);

	NSString *machineName = @(systemInfo.machine);
	NSString *machineNameLine = [NSString stringWithFormat:@"%@ %@", machineName, [UIDevice currentDevice].systemVersion];

	NSString *mailBody = [NSString stringWithFormat:@"\n\n\n\n\n%@\n%@\n%@\n", lineSeparator, appVersionLine, machineNameLine];

	FXDLogObject(mailBody);

	MFMailComposeViewController *emailScene = [[MFMailComposeViewController alloc] initWithNavigationBarClass:nil toolbarClass:nil];
	[emailScene setSubject:subjectLine];
	[emailScene setToRecipients:recipients];
	[emailScene setMessageBody:mailBody isHTML:NO];

	return emailScene;
}

- (MFMailComposeViewController*)emailSceneForSharingImage:(UIImage*)image usingMessage:(NSString*)message {	FXDLog_DEFAULT;

	MFMailComposeViewController *emailScene = [[MFMailComposeViewController alloc] initWithNavigationBarClass:nil toolbarClass:nil];
	[emailScene setSubject:[NSString stringWithFormat:@"[%@]", application_DisplayName]];

	if (image) {
		[emailScene addAttachmentData:UIImageJPEGRepresentation(image, 1.0) mimeType:@"image/jpeg" fileName:@"sharedImage"];
	}

	if (message) {
		[emailScene setMessageBody:message isHTML:NO];
	}

	return emailScene;
}

#pragma mark - Observer

#pragma mark - Delegate
//MARK: MFMailComposeViewControllerDelegate
- (void)mailComposeController:(MFMailComposeViewController*)controller didFinishWithResult:(MFMailComposeResult)result error:(NSError *)error {	FXDLog_DEFAULT;

	FXDLogVariable(result);

	FXDLog_ERROR;

	[controller dismissViewControllerAnimated:YES completion:nil];
}

@end

File: Sources/FXDmoduleMessage.swift


import MessageUI


open class FXDmoduleMessage: NSObject {
	func presentEmailScene(_ emailScene: MFMailComposeViewController?, forPresentingScene presentingScene: UIViewController?, using image: UIImage?, usingMessage message: String?, withRecipients recipients: [String]?) {	fxd_log()

		guard MFMailComposeViewController.canSendMail() else {
			//TODO: alert user
			return
		}


		var emailScene = emailScene
		if emailScene == nil {
			if (image != nil || message != nil) {
				emailScene = self.emailScene(forSharing: image, usingMessage: message)
			}
			else {
				emailScene = self.emailSceneWithMailBody(withRecipients: recipients)
			}
		}

		guard emailScene != nil else {
			//TODO: alert user
			return
		}


		var presentingScene = presentingScene
		if presentingScene == nil {
			presentingScene = UIApplication.shared.mainWindow()?.rootViewController
			fxdPrint(presentingScene as Any)
		}


		emailScene?.mailComposeDelegate = self

		presentingScene?.present(
			emailScene!,
			animated: true,
			completion: nil)
	}

	func emailSceneWithMailBody(withRecipients recipients: [String]?) -> MFMailComposeViewController! {	fxd_log()
#if DEBUG
		fxdPrint("Bundle.main.infoDictionary: ", Bundle.main.infoDictionary)
		fxdPrint("recipients: ", recipients)

		if let currentLanguage = (UserDefaults.standard.object(forKey: "AppleLanguages") as? Array<Any?>)?.first {
			fxdPrint("currentLanguage: ", currentLanguage)
		}
#endif


		let version = Bundle.bundleVersion()
		let displayName = Bundle.bundleDisplayName()

		let subjectLine = "[\(displayName ?? "")]"
		let lineSeparator = "_______________________________"
		let appVersionLine = "\(subjectLine) \(version ?? "")"
		let machineNameCode = UIDevice.machineNameCode()
		let machineNameLine = "\(machineNameCode ?? "") \(UIDevice.current.systemVersion)"
		let mailBody = "\n\n\n\n\n\(lineSeparator)\n\(appVersionLine)\n\(machineNameLine)\n"
		fxdPrint("mailBody: \(mailBody)")


		let emailScene = MFMailComposeViewController(navigationBarClass: nil, toolbarClass: nil)
		emailScene.setSubject(subjectLine)
		emailScene.setToRecipients(recipients ?? [])
		emailScene.setMessageBody(mailBody, isHTML: false)

		return emailScene
	}

	func emailScene(forSharing image: UIImage?, usingMessage message: String?) -> MFMailComposeViewController? {	fxd_log()

		let displayName = Bundle.bundleDisplayName()

		let subjectLine = "[\(displayName ?? "")]"


		let emailScene = MFMailComposeViewController(navigationBarClass: nil, toolbarClass: nil)
		emailScene.setSubject(subjectLine)

		if let jpegData = image?.jpegData(compressionQuality: 1.0) {
			emailScene.addAttachmentData(jpegData, mimeType: "image/jpeg", fileName: "sharedImage")
		}

		if message != nil {
			emailScene.setMessageBody(message!, isHTML: false)
		}

		return emailScene
	}
}


extension FXDmoduleMessage: MFMailComposeViewControllerDelegate {
	public func mailComposeController(_ controller: MFMailComposeViewController, didFinishWith result: MFMailComposeResult, error: Error?) {	fxd_log()

		fxdPrint(result)
		fxdPrint(error)

		controller.dismiss(animated: true, completion: nil)
	}
}
